### PR TITLE
Fix wallet_scanQRCode rpc method

### LINKED ISSUE #1485: scanQRCode seems to be missing
**Describe the bug**
`scanQRCode` api to enable QR code scanning seems to have been removed. It cannot be found anymore

**Screenshots**
Sent some screenshots through TestFlight

**Smartphone (please complete the following information):**
 - Device: iPhone11 Pro
 - OS: ios13

-------------------------------------------------------------
_to be added after bug submission by internal support / PM_
**Severity**

- How critical is the impact of this bug on a user?
- Add stats if available on % of customers impacted
- Is this visible to all users?
- Is this tech debt?

### PR DESCRIPTION
**Pull request by EtDu**
**Description**

Calling `provider.request({ method: 'wallet_scanQRCode' })` was returning the error:
'RPC Error: Error: JsonRpcEngine - response has no error or result for request:'
because it wouldn't wait for the QR code to be scanned before passing the response as empty. Now it properly waits for the user to scan before passing the given result. 

This also adds the option for a regex parameter outlined in https://github.com/ethereum/EIPs/issues/945 to enable arbitrary (non ETH address) QR code scanning.

Will reply and close #1485 after this is merged.

**Issue**

https://trello.com/c/J58J4xOC/77-bug-scan-qr-code-missing

### CODE REVIEW

File: app/components/Views/BrowserTab/index.js [hunk 1]
```diff
@@ -637,19 +637,28 @@ export class BrowserTab extends PureComponent {
...
 
 				wallet_scanQRCode: async () => {
-					this.props.navigation.navigate('QRScanner', {
-						onScanSuccess: data => {
-							let result = data;
-							if (data.target_address) {
-								result = data.target_address;
-							} else if (data.scheme) {
-								result = JSON.stringify(data);
+					await new Promise((resolve, reject) => {
```
Comment (andrepimenta) [inline app/components/Views/BrowserTab/index.js:640]: Here I think you can simplify either:
`wallet_scanQRCode: () => new Promise((resolve, reject) => { ... }) `

Or even simpler:
async () => {

...
 throw new Error("This is an error") // Same as reject({message: "This is an error"})
...

 return // Same as resolve()
}

Comment (EtDu) [reply app/components/Views/BrowserTab/index.js]: Good catch

File: app/components/Views/BrowserTab/index.js [hunk 2]
```diff
@@ -636,22 +636,30 @@ export class BrowserTab extends PureComponent {
...
-								result = JSON.stringify(data);
+				wallet_scanQRCode: () =>
+					new Promise((resolve, reject) => {
+						this.props.navigation.navigate('QRScanner', {
+							onScanSuccess: data => {
+								const regex = new RegExp(req.params[0]);
+								if (regex && !regex.exec(data)) {
+									reject({ message: 'NO_REGEX_MATCH', data });
+								} else if (!regex && !/^(0x){1}[0-9a-fA-F]{40}$/i.exec(data.target_address)) {
+									reject({ message: 'INVALID_ETHEREUM_ADDRESS', data: data.target_address });
```
Comment (estebanmino) [inline app/components/Views/BrowserTab/index.js:647]: are these error names standard? I couldn't find these on the link you posted

Comment (EtDu) [reply app/components/Views/BrowserTab/index.js:647]: This PR doesn't technically implement the EIP as it's outlined anyway, but there doesn't appear to be a standard for the error names

**Comment by ibrahimtaveras00** [comment]
@EtDu I went ahead and followed the instructions to reproduce [here](https://github.com/MetaMask/metamask-mobile/issues/1485#issuecomment-614349802) and I'm still seeing that it's not detected via the link https://cipher-qr-scan.netlify.com/

Screen Shot 2020-09-16 at 12 11 27 PM

**Comment by EtDu** [comment]
@ibrahimtaveras00 
It's not meant to work with the Cipher API. This implementation defines the RPC method `wallet_scanQRCode` that dApps must access via `provider.request()`. NOT via `provider.scanQRCode`. Kyberswap relies on webRTC support (which iOS webview doesn't have). They would need to change their code to work with this implementation if they wanted their scan functionality to work with MetaMask.

To test this, open a webview debug console and type `ethereum.request({ method: 'wallet_scanQRCode' })`

**Comment by nicc938** [comment]
i come back too hame  make it my pc

ในวันที่ พฤ. 17 ก.ย. 2020 เวลา 10:33 Etienne Dusseault <
notifications@github.com> เขียนว่า:

**Comment by nicc938** [comment]
sryy


ในวันที่ พฤ. 17 ก.ย. 2020 เวลา 10:43 Nicc Lee <juab.leela@gmail.com>
เขียนว่า:

**Comment by ibrahimtaveras00** [comment]
@EtDu I get this when I tap X to close the scanner on browser = http://recordit.co/TLdR6OWJxX

Screen Shot 2020-09-18 at 8 52 07 PM

Same on iOS = http://recordit.co/8IbXZ81Phe

**Comment by EtDu** [comment]
@ibrahimtaveras00 That's an internal error that does not appear in production and has existed before this PR.

**Review by ibrahimtaveras00 — APPROVED** [review]
👍

**Pull request merged**